packages/eslint-plugin/src/rules/prefer-optional-chain-utils/analyzeChain.ts

```diff
@@ -0,0 +1,552 @@
+          },
+        ];
+      }
+
+      case AST_NODE_TYPES.MemberExpression: {
+        const propertyText = sourceCode.getText(node.property);
+        return [
+          ...flattenChainExpression(sourceCode, node.object),
+          {
+            nonNull: node.object.type === AST_NODE_TYPES.TSNonNullExpression,
+            optional: node.optional,
+            precedence: node.computed
+              ? // computed is already wrapped in [] so no need to wrap in () as well
+                util.OperatorPrecedence.Invalid
+              : util.getOperatorPrecedenceForNode(node.property),
+            requiresDot: !node.computed,
+            text: node.computed ? `[${propertyText}]` : propertyText,
+          },
+        ];
+      }
+
+      case AST_NODE_TYPES.TSNonNullExpression:
+        return flattenChainExpression(sourceCode, node.expression);
+
+      default:
+        return [
+          {
+            nonNull: false,
+            optional: false,
+            precedence: util.getOperatorPrecedenceForNode(node),
+            requiresDot: false,
+            text: sourceCode.getText(node),
+          },
+        ];
+    }
+  }
+}
+
+export function analyzeChain(
+  context: RuleContext<
+    PreferOptionalChainMessageIds,
+    [PreferOptionalChainOptions]
+  >,
+  sourceCode: SourceCode,
+  parserServices: ParserServicesWithTypeInformation,
+  options: PreferOptionalChainOptions,
+  operator: TSESTree.LogicalExpression['operator'],
+  chain: ValidOperand[],
+): void {
+  // need at least 2 operands in a chain for it to be a chain
+  if (
+    chain.length <= 1 ||
+    /* istanbul ignore next -- previous checks make this unreachable, but keep it for exhaustiveness check */
+    operator === '??'
+  ) {
+    return;
+  }
+
+  const analyzeOperand = (() => {
+    switch (operator) {
+      case '&&':
+        return analyzeAndChainOperand;
+
+      case '||':
+        return analyzeOrChainOperand;
+    }
+  })();
+
+  let subChain: ValidOperand[] = [];
+  const maybeReportThenReset = (
+    newChainSeed?: readonly ValidOperand[],
+  ): void => {
+    if (subChain.length > 1) {
+      context.report({
+        messageId: 'preferOptionalChain',
+        loc: {
+          start: subChain[0].node.loc.start,
+          end: subChain[subChain.length - 1].node.loc.end,
+        },
+        ...getFixer(sourceCode, parserServices, operator, options, subChain),
+      });
+    }
+
+    // we've reached the end of a chain of logical expressions
+    // we know the validated
+    subChain = newChainSeed ? [...newChainSeed] : [];
+  };
+
+  for (let i = 0; i < chain.length; i += 1) {
+    const lastOperand = subChain[subChain.length - 1] as
+      | ValidOperand
+      | undefined;
+    const operand = chain[i];
+
+    const validatedOperands = analyzeOperand(parserServices, operand, i, chain);
+    if (!validatedOperands) {
+      // TODO - check if the name is a superset/equal - if it is, then it was
```

Comment:
> TODO

Is there a tracking issue? _(here and elsewhere)_

Reply:
tracking issues are for organised people.
Filed #7170